Support properties with setter type different from getter type

ilevkivskyi opened the pull request:
Fixes https://github.com/python/mypy/issues/3004
Fixes https://github.com/python/mypy/issues/11892
Fixes https://github.com/python/mypy/issues/12892
Fixes https://github.com/python/mypy/issues/14301

_Note:_ this PR should be reviewed with "hide whitespace" option (in couple long functions I replace huge `if x: ...` with `if not x: return; ...` to reduce indent level).

The core logic is quite straightforward (almost trivial). The only couple things are:
* We should be careful with binder (we simpy can't use it for properties with setter type different from getter type, since we don't know underlying property implementation)
* We need to handle gracefully existing settable properties that are generated by plugins

The tricky part is subclassing and protocols. The summary is as following:
* For protocols I simply implement everything the "correct way", i.e. for settable attributes (whether it is a variable or a settable property) compare getter types covariantly and setter types contravariantly. The tricky part here is generating meaningful error messages that are also not too verbose.
* For subclassing I cannot simply do the same, because there is a flag about covariant mutable override, that is off by default. So instead what I do is if either subclass node, or superclass node is a "custom property" (i.e. a property with setter type different from getter type), then I use the "correct way", otherwise the old logic (i.e. flag dependent check) is used.

Two things that are not implemented are multiple inheritance, and new generic syntax (inferred variance). In these cases setter types are simply ignored. There is nothing conceptually difficult about these, I simply run out of steam (and the PR is getting big). I left `TODO`s in code for these. In most cases these will generate false negatives (and they are already kind of corner cases) so I think it is OK to postpone these indefinitely.

ilevkivskyi commented:
OK, I think the primer looks very good. There are few new errors (for example in `kornia`), it looks like all of them are because we don't use binder anymore, and as I explained in PR description IMO this is correct.

The weird errors in pandas are because of this:
```python
HashableT = TypeVar("HashableT", bound=Hashable)

class DataFrame:  # not generic in HashableT
    @property
    def columns(self) -> Index[str]: ...
    @columns.setter
    def columns(
        self, cols: AnyArrayLike | list[HashableT] | tuple[HashableT, ...]
    ) -> None: ...
```
What would this even mean, why not simply use `Hashable`? One possible way I see is to give error at definition if there are type variables not bound by `self` and erase them. But this would be a bunch of work for a very weird error. Any opinions? @JukkaL

ilevkivskyi commented:
Oh, I think I understand what they want (probably):
```python
HashableT = TypeVar("HashableT", bound=Hashable)
def foo(x: list[HashableT]) -> None: ...
def bar(x: list[Hashable]) -> None: ...

x: list[str]
foo(x)  # OK
bar(x)  # Invalid because invariance
```
This is a clever trick, but why not simply use `Sequence`? Also supporting the pattern in the above comment for this purpose would be a pain.

A5rocks commented:
Some relevant issues (found using https://github.com/A5rocks/mypy-issues):
 - https://github.com/python/mypy/issues/14301 gets fixed
 - https://github.com/python/mypy/issues/12892 gets fixed
 - https://github.com/python/mypy/issues/11892 gets fixed

It looks like this messes up property deleters, though. Like this (taken from https://github.com/python/mypy/issues/17928):
```py
class C:
    @property
    def x(self) -> int:
        return 0

    @x.deleter
    def x(self) -> None:
        pass
```

JukkaL commented on mypy/nodes.py:
```diff
@@ -1076,6 +1079,7 @@ def serialize(self) -> JsonDict:
             "name": self._name,
             "fullname": self._fullname,
             "type": None if self.type is None else self.type.serialize(),
+            "setter_type": None if self.setter_type is None else self.setter_type.serialize(),
```
Maybe add a serialization test case?

JukkaL commented on test-data/unit/check-protocols.test:
```diff
@@ -4246,3 +4246,219 @@ class SupportsAdd(Protocol):
+        # N: Following member(s) of "A4" have conflicts: \
+        # N:     foo: expected "B1", got "str" \
+        # N:     foo: expected setter type "C1", got "str"
+[builtins fixtures/property.pyi]
```
Does mypy daemon need to be modified to detect changed in setter type, i.e. does `astdiff` need to be updated? Maybe add a daemon test case that modifies the setter type.

ilevkivskyi replied:
Yes, good catch!

JukkaL commented on test-data/unit/check-classes.test:
```diff
@@ -8184,3 +8395,18 @@ class C:
         __qualname__  # E: Name "__qualname__" is not defined
+
+[case testPropertySetterType]
+class A:
```
Also check a generic class with a property that uses a type var type, and a subclass that uses non-trivial generic inheritance?

ilevkivskyi replied:
There is `testOverridePropertyGeneric` for overrides, so I assume here you mean test _access_ on non-trivial subclass.

JukkaL reviewed (approved):
Thanks for the PR! This is a long-requested feature and many users will be happy.

Looks good, just some ideas about further tests. And also please look into comments by others before merging.

ilevkivskyi commented:
@A5rocks @JukkaL Thanks for comments! I think I implemented all of them (and updated the PR description). I will decide tomorrow about what to do with the weird `pandas` corner case I mentioned in mt first comment, and then merge the PR if there are no further comments.

ilevkivskyi commented:
After some thinking I am going to merge this as is. Then if people will complain about the pattern in pandas we can do something, because at this moment it is hard to tell how important it is, and I don't want to guess. It may be a super-rare use case, and I don't want to spend time implementing support for this weird pattern unless it is actually important for some people.

hauntsaninja commented:
This is going to make a lot of people very happy. Tested out a few things and looks great, thank you for working on it

ilevkivskyi commented:
@Dr-Irv Great, thanks for info! If you are switching to plain `Hashable`, I would also recommend to use `Sequence` instead of `list | tuple` there to avoid false positives in certain cases (unless pandas actually requires only lists or tuples for some reason).

Dr-Irv commented:
> @Dr-Irv Great, thanks for info! If you are switching to plain `Hashable`, I would also recommend to use `Sequence` instead of `list | tuple` there to avoid false positives in certain cases (unless pandas actually requires only lists or tuples for some reason).

We can't use `Sequence` because we don't want to accept `str` there.
